Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -1424,6 +1460,115 @@ TEST_F(SerializerTest, EncodesResumeTokens) {
   ExpectRoundTrip(model, proto);
 }
 
+TEST_F(SerializerTest, EncodesListenRequestLabels) {
+  core::Query q = Query("docs");
+
+  std::vector<
+      std::pair<QueryPurpose, std::unordered_map<std::string, std::string>>>
+      purposeToLabel = {
+          {QueryPurpose::Listen, {}},
+          {QueryPurpose::LimboResolution,
+           {{"goog-listen-tags", "limbo-document"}}},
+          {QueryPurpose::ExistenceFilterMismatch,
+           {{"goog-listen-tags", "existence-filter-mismatch"}}},
+      };
+
+  for (const auto& p : purposeToLabel) {
+    QueryData model(q, 1, 0, p.first, SnapshotVersion::None(), Bytes(1, 2, 3));
```

Comment:
Note: there's an overload of `QueryData` constructor that takes fewer arguments, you can omit everything after `p.first`.